cpp/arcticdb/column_store/column.hpp

```diff
@@ -646,6 +646,72 @@ class Column {
         return *res;
     }
 
+    template<typename input_tdt>
+    static void for_each(const Column& input_column,
```

Comment:
I don't think we need `std::funciton` here. We can add `typename FunctorT&&` template parameter as we don't store the function in memory. `std::function` uses type erasure and can sometimes dynamically allocate memory and have additional indirection.

Reply:
I like using this format, as it makes it explicit that you cannot pass in arbitrary functions, they must have a very specific format.
As a general rule, `std::function` will avoid the heap if the function it is wrapping is small enough, although there is no way to tell if this is the case.
`folly::Function`, however, has a `heapAllocatedMemory()` method. I'm currently running all the tests with an `internal::check` that the heap allocated memory is always zero. If this passes on all platforms, we can remove the checks and know that the existing uses do no heap allocation. The tricky thing is enforcing this going forwards.

Reply:
Concepts will help this case when we enable C++20. You can achieve the same thing with enable_if and company, but it's kind of ugly.